<<<<<<< HEAD
---
description: 'Disallow extra non-null assertions.'
---

> ðŸ›‘ This file is source code, not the primary documentation location! ðŸ›‘
>
> See **https://typescript-eslint.io/rules/no-extra-non-null-assertion** for documentation.
=======
# Disallow extra non-null assertion (`no-extra-non-null-assertion`)
>>>>>>> 92030a1a5b086ddccf899c6bd734063c39dae48d

The `!` non-null assertion operator in TypeScript is used to assert that a value's type does not include `null` or `undefined`.
Using the operator any more than once on a single value does nothing.

<<<<<<< HEAD
## Examples

<!--tabs-->

### âŒ Incorrect
=======
Examples of **incorrect** code for this rule:
>>>>>>> 92030a1a5b086ddccf899c6bd734063c39dae48d

```ts
const foo: { bar: number } | null = null;
const bar = foo!!!.bar;
```

```ts
function foo(bar: number | undefined) {
  const bar: number = bar!!!;
}
```

```ts
function foo(bar?: { n: number }) {
  return bar!?.n;
}
```

Examples of **correct** code for this rule:

```ts
const foo: { bar: number } | null = null;
const bar = foo!.bar;
```

```ts
function foo(bar: number | undefined) {
  const bar: number = bar!;
}
```

```ts
function foo(bar?: { n: number }) {
  return bar?.n;
}
```

## How to use

```json
{
  "@typescript-eslint/no-extra-non-null-assertion": ["error"]
}
```
