# Disallows using a non-null assertion after an optional chain expression (`no-non-null-asserted-optional-chain`)

`?.` optional chain expressions provide `undefined` if an object is `null` or `undefined`.
Using a `!` non-null assertion to assert the result of an `?.` optional chain expression is non-nullable is likely wrong.

> Most of the time, either the object was not nullable and did not need the `?.` for its property lookup, or the `!` is incorrect and introducing a type safety hole.

<<<<<<< HEAD
## Examples

<!--tabs-->

### âŒ Incorrect
=======
Examples of **incorrect** code for this rule:
>>>>>>> 92030a1a5b086ddccf899c6bd734063c39dae48d

```ts
foo?.bar!;
foo?.bar()!;
```

Examples of **correct** code for this rule:

```ts
foo?.bar;
foo?.bar();
```

## Further Reading

- [TypeScript 3.7 Release Notes](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html)
- [Optional Chaining Proposal](https://github.com/tc39/proposal-optional-chaining/)
