# Disallows invocation of `require()` (`no-require-imports`)

Prefer the newer ES6-style imports over `require()`.

<<<<<<< HEAD
## Examples

<!--tabs-->

### ❌ Incorrect
=======
## Rule Details

Examples of **incorrect** code for this rule:
>>>>>>> 92030a1a5b086ddccf899c6bd734063c39dae48d

```ts
const lib1 = require('lib1');
const { lib2 } = require('lib2');
import lib3 = require('lib3');
```

Examples of **correct** code for this rule:

```ts
import * as lib1 from 'lib1';
import { lib2 } from 'lib2';
import * as lib3 from 'lib3';
```

## When Not To Use It

<<<<<<< HEAD
If you don't care about using newer module syntax, then you will not need this rule.

## Related To

- [`no-var-requires`](./no-var-requires.md)
=======
If you don't care about TypeScript module syntax, then you will not need this rule.

## Compatibility

- TSLint: [no-require-imports](https://palantir.github.io/tslint/rules/no-require-imports/)
>>>>>>> 92030a1a5b086ddccf899c6bd734063c39dae48d
