# Warns for any two overloads that could be unified into one by using a union or an optional/rest parameter (`unified-signatures`)

Warns for any two overloads that could be unified into one by using a union or an optional/rest parameter.

Function overload signatures are a TypeScript way to define a function that can be called in multiple very different ways.
Overload signatures add syntax and theoretical bloat, so it's generally best to avoid using them when possible.
Switching to union types and/or optional or rest parameters can often avoid the need for overload signatures.

This rule reports when function overload signatures can be replaced by a single function signature.

<<<<<<< HEAD
## Examples

<!--tabs-->

### ❌ Incorrect
=======
Examples of **incorrect** code for this rule:
>>>>>>> 92030a1a5b086ddccf899c6bd734063c39dae48d

```ts
function f(x: number): void;
function f(x: string): void;
```

```ts
f(): void;
f(...x: number[]): void;
```

Examples of **correct** code for this rule:

```ts
function f(x: number | string): void;
```

```ts
function f(x?: ...number[]): void;
```

<<<<<<< HEAD
```ts
// This rule won't check overload signatures with different rest parameter types.
// See https://github.com/microsoft/TypeScript/issues/5077
function f(...a: number[]): void;
function f(...a: string[]): void;
```

## Options

### `ignoreDifferentlyNamedParameters`

Examples of code for this rule with `ignoreDifferentlyNamedParameters`:

<!--tabs-->

### ❌ Incorrect

```ts
function f(a: number): void;
function f(a: string): void;
```

### ✅ Correct

```ts
function f(a: number): void;
function f(b: string): void;
```

## Options
=======
## Related to

- TSLint: [`unified-signatures`](https://palantir.github.io/tslint/rules/unified-signatures/)
>>>>>>> 92030a1a5b086ddccf899c6bd734063c39dae48d
